library/modules/Units.cpp

```diff
@@ -644,10 +644,9 @@ bool Units::isDanger(df::unit *unit) {
 
     return isCrazed(unit)
         || isInvader(unit)
-        || isOpposedToLife(unit)
         || isAgitated(unit)
         || isSemiMegabeast(unit)
-        || isNightCreature(unit)
+        || (!isVisiting(unit) && (isOpposedToLife(unit) || isNightCreature(unit)))
```

Comment:
isVisiting also returns true for "unwanted" visitors. Is this check enough?

Reply:
Actually, you're right, forgotton beasts should be marked as visitor_uninvited = true. So they would be dangerous.

I'm confused why the isVisitor() function permits a hostile as a visitor? The tests in that function imply it's a friendly visit. IMO, that should go into the isDanger() or similar function

Reply:
Sorry, must've been thinking of another function. Thought there were more tests in the isVisitor() function.

I suppose then just check for visitor flag for that then.  But if we want to allow for i.e. friendly demons then I think that visitor flag check would need to be done at the top before the rest of the checks.